cmd/clusterctl/client/alpha/rollout_rollbacker.go

```diff
@@ -0,0 +1,171 @@
+func (r *rollout) ObjectRollbacker(proxy cluster.Proxy, tuple util.ResourceTuple, namespace string, toRevision int64) error {
+	switch tuple.Resource {
+	case machineDeployment:
+		deployment, err := getMachineDeployment(proxy, tuple.Name, namespace)
+		if err != nil || deployment == nil {
+			return errors.Wrapf(err, "failed to fetch %v/%v", tuple.Resource, tuple.Name)
+		}
+		if deployment.Spec.Paused {
+			return errors.Errorf("can't rollback a paused MachineDeployment (run rollout resume first): %v/%v\n", tuple.Resource, tuple.Name)
+		}
+		if err := rollbackMachineDeployment(proxy, deployment, tuple.Name, namespace, toRevision); err != nil {
+			return err
+		}
+	default:
+		return errors.Errorf("invalid resource type %q, valid values are %v", tuple.Resource, validResourceTypes)
+	}
+	return nil
+}
+
+// rollbackMachineDeployment will rollback to a previous MachineSet revision used by this MachineDeployment.
+func rollbackMachineDeployment(proxy cluster.Proxy, d *clusterv1.MachineDeployment, name, namespace string, toRevision int64) error {
+	log := logf.Log
+	c, err := proxy.NewClient()
+	if err != nil {
+		return err
+	}
+
+	if toRevision < 0 {
+		return errors.Errorf("revision number cannot be negative: %v", toRevision)
+	}
+	msList, err := getMachineSetsForDeployment(proxy, d)
+	if err != nil {
+		return err
+	}
+	log.V(7).Info("Found MachineSets", "count", len(msList))
+	msForRevision, err := findMachineDeploymentRevision(toRevision, msList)
+	if err != nil {
+		return err
+	}
+	log.V(7).Info("Found revision", "revision", msForRevision)
+	patchHelper, err := patch.NewHelper(d, c)
+	if err != nil {
+		return err
+	}
+	// Copy template into the machinedeployment (excluding the hash)
+	revMSTemplate := *msForRevision.Spec.Template.DeepCopy()
+	delete(revMSTemplate.Labels, mdutil.DefaultMachineDeploymentUniqueLabelKey)
+
+	d.Spec.Template = revMSTemplate
+	return patchHelper.Patch(context.TODO(), d)
+}
+
+// findMachineDeploymentRevision finds the specific revision in the machine sets
+func findMachineDeploymentRevision(toRevision int64, allMSs []*clusterv1.MachineSet) (*clusterv1.MachineSet, error) {
+
+	var (
+		latestMachineSet   *clusterv1.MachineSet
+		latestRevision     = int64(-1)
+		previousMachineSet *clusterv1.MachineSet
+		previousRevision   = int64(-1)
+	)
+	for _, ms := range allMSs {
+		if v, err := mdutil.Revision(ms); err == nil {
+			if toRevision == 0 {
+				if latestRevision < v {
+					// newest one we've seen so far
+					previousRevision = latestRevision
+					previousMachineSet = latestMachineSet
+					latestRevision = v
+					latestMachineSet = ms
+				} else if previousRevision < v {
+					// second newest one we've seen so far
+					previousRevision = v
+					previousMachineSet = ms
+				}
+			} else if toRevision == v {
+				return ms, nil
+			}
+		}
+	}
+
+	if toRevision > 0 {
+		return nil, errors.Errorf("unable to find specified revision: %v", toRevision)
+	}
```

Comment:
I'm guessing this is added to provide a better error msg because I see that the next check would also result in returning an error.

Reply:
Yes, the specific revision desired was not found. It's different from the next error which is that there is no previous MS associated with this deployment.